llama_index/prompts/base.py

```diff
@@ -1,173 +1,295 @@
-"""Base module for prompts."""
+"""Prompts."""
+
+
+from abc import ABC, abstractmethod
 from copy import deepcopy
-from typing import Any, Dict, List, Optional
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+from pydantic import BaseModel
 
-from llama_index.bridge.langchain import BasePromptTemplate as BaseLangchainPrompt
-from llama_index.bridge.langchain import PromptTemplate as LangchainPrompt
+from llama_index.bridge.langchain import BasePromptTemplate as LangchainTemplate
+from llama_index.bridge.langchain import ConditionalPromptSelector as LangchainSelector
 from llama_index.llms.base import LLM, ChatMessage
+from llama_index.llms.generic_utils import messages_to_prompt, prompt_to_messages
+from llama_index.llms.langchain import LangChainLLM
 from llama_index.llms.langchain_utils import from_lc_messages
-from llama_index.prompts.prompt_selector import PromptSelector
 from llama_index.prompts.prompt_type import PromptType
+from llama_index.prompts.utils import get_template_vars
 from llama_index.types import BaseOutputParser
 
 
-class Prompt:
-    """Prompt class for LlamaIndex.
+class BasePromptTemplate(BaseModel, ABC):
+    metadata: Dict[str, Any]
+    template_vars: List[str]
+    kwargs: Dict[str, str]
+    output_parser: Optional[BaseOutputParser]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    @abstractmethod
+    def partial_format(self, **kwargs: Any) -> "BasePromptTemplate":
+        ...
+
+    @abstractmethod
+    def format(self, llm: Optional[LLM] = None, **kwargs: Any) -> str:
+        ...
+
+    @abstractmethod
+    def format_messages(
+        self, llm: Optional[LLM] = None, **kwargs: Any
+    ) -> List[ChatMessage]:
+        ...
+
+
+class PromptTemplate(BasePromptTemplate):
```

Comment:
Just trying to think about how we can mitigate disruption for users. tbh prompt customization is pretty common for the power users.

From my understanding, old completion prompts won't be affected (i.e. `Prompt("text")` or `QuestionAnswerPrompt("text")`)

What WILL be changed is the flow where we had

```
CHAT_TEXT_QA_PROMPT_LC = ChatPromptTemplate.from_messages(TEXT_QA_PROMPT_TMPL_MSGS)
CHAT_TEXT_QA_PROMPT = QuestionAnswerPrompt.from_langchain_prompt(CHAT_TEXT_QA_PROMPT_LC)
```

Is there any way we can sneak in a `from_langchain_prompt` method somewhere? Tbh looking at the code, I don't think so (so just flagging that this specific use-case will break)

Reply:
Thanks for flagging this. You are right.

If the user is using the old syntax for building chat templates (or selector templates), this PR will break that.
Since we are using separate classes (instead of a monolithic prompt class) to handle plain, chat, selector, langchain-compat versions of templates, it's hard to continue  supporting the old `from_langchain_prompt`.